### PR TITLE
fix: prevent infinite recursion in getDirtyFields with circular references

### PR DESCRIPTION
**Pull request by kamja44**
Added circular reference detection using WeakSet to prevent stack overflow when form values contain self-referencing objects or arrays.

Changes:
- Added 'visited' parameter (WeakSet) to markFieldsDirty function
- Added 'visited' parameter (WeakSet) to getDirtyFields function
- Track visited objects to detect circular references
- Return early when encountering already-visited objects
- Added test cases for circular reference scenarios

Fixes potential RangeError: Maximum call stack size exceeded when processing form data with circular object/array references.

### CODE REVIEW

File: src/logic/getDirtyFields.ts [hunk 1]
```diff
@@ -31,9 +42,16 @@ export default function getDirtyFields<T>(
...
+  if (typeof data === 'object' && data !== null) {
+    if (visited.has(data)) {
+      return dirtyFieldsFromValues || {};
+    }
+    visited.add(data);
+  }
   if (!dirtyFieldsFromValues) {
-    dirtyFieldsFromValues = markFieldsDirty(formValues);
+    dirtyFieldsFromValues = markFieldsDirty(formValues, {}, visited);
   }
```
[the remark this answers is not in the document]
Comment (kamja44) [reply src/logic/getDirtyFields.ts:55]: I fixed this by treating `visited` as a call stack tracker instead of a global set.

The key changes:
- Line 54: `markFieldsDirty(formValues)` gets its own `visited` set
- Line 86: `visited.delete(data)` removes objects after processing

This way circular references are caught (object in current call stack) but shared objects at different paths get compared normally (already removed from the set).

Tests added at lines 446-466 and 468-480 cover these cases.

File: src/__tests__/logic/getDirtyFields.test.ts [hunk 2]
```diff
@@ -376,4 +376,45 @@ describe('getDirtyFields', () => {
...
+
+    // Should not throw stack overflow error with circular references
+    expect(() => {
+      getDirtyFields(circularObj, formValues);
+    }).not.toThrow();
+
+    const result: any = getDirtyFields(circularObj, formValues);
+    expect(result).toHaveProperty('name');
+    expect(result.name).toBe(true); // Different value, so dirty
+  });
```
[the remark this answers is not in the document]
Comment (kamja44) [reply src/__tests__/logic/getDirtyFields.test.ts:395]: Added! See lines 395-397 - now verifying the full nested structure and that circular references return empty objects/arrays as expected.

File: src/__tests__/logic/getDirtyFields.test.ts [hunk 3]
```diff
@@ -376,4 +376,45 @@ describe('getDirtyFields', () => {
...
+  it('should handle self-referencing arrays without infinite recursion', () => {
+    const circularArray: any = [{ id: 1 }];
+    circularArray.push(circularArray); // Array contains itself
+
+    const formArray: any = [{ id: 2 }];
+    formArray.push(formArray);
+
+    expect(() => {
+      getDirtyFields({ items: circularArray }, { items: formArray });
+    }).not.toThrow();
```
[the remark this answers is not in the document]
Comment (kamja44) [reply src/__tests__/logic/getDirtyFields.test.ts:406]: added assertions at lines 416-422 to verify the array structure and element comparisons. The test now checks that the result has the 'items' property, verifies it's an array, validates the first element's 'id' property is marked as dirty (true), and confirms the circular reference at index 1 returns an empty array.

File: src/__tests__/logic/getDirtyFields.test.ts [hunk 4]
```diff
@@ -376,4 +376,45 @@ describe('getDirtyFields', () => {
...
+  it('should handle deeply nested circular references', () => {
+    const obj1: any = { a: { b: { c: {} } } };
+    obj1.a.b.c.loop = obj1.a;
+
+    const obj2: any = { a: { b: { c: {} } } };
+    obj2.a.b.c.loop = obj2.a;
+
+    expect(() => {
+      getDirtyFields(obj1, obj2);
+    }).not.toThrow();
```
[the remark this answers is not in the document]
Comment (kamja44) [reply src/__tests__/logic/getDirtyFields.test.ts:418]: added assertions at lines 438-443 to validate the output structure. The test now verifies that the result correctly traverses the nested structure (a → b → c → loop), ensures all non-circular properties are properly checked, and confirms the circular reference at 'loop' returns an empty object instead of causing infinite recursion.

File: src/logic/getDirtyFields.ts [hunk 5]
```diff
@@ -44,12 +62,14 @@ export default function getDirtyFields<T>(
...
           Array.isArray(value) ? [] : {},
+          visited,
         );
       } else {
         getDirtyFields(
           value,
           isNullOrUndefined(formValues) ? {} : formValues[key],
           dirtyFieldsFromValues[key],
+          visited,
         );
```
[the remark this answers is not in the document]
Comment (kamja44) [reply src/logic/getDirtyFields.ts:73]: This was indeed an issue in the initial implementation. I've fixed it by using a call-stack based tracking approach (lines 57-60 and 85-87). The key change is adding `visited.delete(data)` at line 86, which removes the object from the visited set when exiting the recursion call.

This allows the same object at different paths to be processed correctly while still preventing infinite loops from circular references. The test at lines 446-466 specifically validates this scenario - when `shared` object appears at both `data.a` and `data.b`, it's properly compared at each path:
- `data.a` (shared) vs `formValues.a` ({ value: 'other' }) → dirty
- `data.b` (shared) vs `formValues.b` (shared) → not dirty

The visited set now tracks only the current call stack, not the entire traversal history.

File: src/__tests__/logic/getDirtyFields.test.ts [hunk 6]
```diff
@@ -376,4 +376,45 @@ describe('getDirtyFields', () => {
...
+  });
```
[the remark this answers is not in the document]
Comment (kamja44) [reply src/__tests__/logic/getDirtyFields.test.ts:419]: I've already added this test case at lines 468-480. The test validates exactly this cross-tree shared reference scenario:

- `data.a` (sharedObj) vs `formValues.a` ({ value: 'other' }) → correctly detects as dirty
- `data.b` ({ value: 'different' }) vs `formValues.b` (sharedObj) → correctly detects as dirty

The call-stack based tracking (with `visited.delete` at line 86) ensures that shared objects between `data` and `formValues` are properly compared at each path, rather than being skipped due to being in the visited set.

Additionally, lines 446-466 test shared references within the same tree (data.a and data.b both reference the same object), which is another important edge case.

**Comment by kamja44** [comment]
@bluebill1049 
Hi! The bundlewatch check is failing because the bundle size increased from 11.1KB to 11.11KB (gzip).

This small increase (+0.01KB) is due to adding WeakSet-based circular reference tracking to fix the infinite recursion bug. The implementation uses call-stack tracking (add on entry, delete on exit) which is already minimal.

Can I update the maxSize limit from 11.1KB to 11.2KB in package.json to accommodate this fix?

**Comment by kamja44** [comment]
@bluebill1049 
Run tests are passing now (related to changes in #13244).
This PR is ready for review. Could you please take a look?
